packages/beacon-node/src/util/bufferPool.ts

```diff
@@ -0,0 +1,77 @@
+import {Metrics} from "../metrics/metrics.js";
+
+/**
+ * If consumer wants more memory than available, we grow the buffer by this ratio.
+ */
+const GROW_RATIO = 1.1;
+
+/**
+ * A simple implementation to manage a single buffer.
+ * This is initially used for state serialization at every epoch and for state reload.
+ * We can enhance and use this for other purposes in the future.
+ */
+export class BufferPool {
+  private buffer: Uint8Array;
+  private inUse = false;
+  private currentKey: number;
+  private readonly metrics: Metrics["bufferPool"] | null = null;
+
+  constructor(size: number, metrics: Metrics | null = null) {
+    this.buffer = new Uint8Array(Math.floor(size * GROW_RATIO));
+    this.currentKey = 0;
+    if (metrics) {
+      this.metrics = metrics.bufferPool;
+      metrics.bufferPool.length.addCollect(() => {
+        metrics.bufferPool.length.set(this.buffer.length);
+      });
+    }
+  }
+
+  get length(): number {
+    return this.buffer.length;
+  }
+
+  /**
+   * Returns a buffer of the given size with all 0.
+   * If the buffer is already in use, return null.
+   * Grow the buffer if the requested size is larger than the current buffer.
+   */
+  alloc(size: number): {buffer: Uint8Array; key: number} | null {
+    return this.doAlloc(size, false);
+  }
+
+  /**
+   * Same to alloc() but the buffer is not zeroed.
+   */
+  allocUnsafe(size: number): {buffer: Uint8Array; key: number} | null {
+    return this.doAlloc(size, true);
+  }
+
+  private doAlloc(size: number, isUnsafe = false): {buffer: Uint8Array; key: number} | null {
+    if (this.inUse) {
+      this.metrics?.misses.inc();
+      return null;
+    }
+    this.inUse = true;
+    this.metrics?.hits.inc();
+    this.currentKey += 1;
+    if (size > this.buffer.length) {
+      this.metrics?.grows.inc();
+      this.buffer = new Uint8Array(Math.floor(size * GROW_RATIO));
+    }
+    const bytes = this.buffer.subarray(0, size);
+    if (!isUnsafe) {
+      bytes.fill(0);
+    }
+    return {buffer: bytes, key: this.currentKey};
+  }
+
+  /**
+   * Marks the buffer as free.
+   */
+  free(key: number): void {
```

Comment:
I was wondering why we need to track a `currentKey` at all? But I guess we need to handle cases where two calls to `getOrReload` can happen at the same time?

Reply:
yes at least now we call it in 2 flows: to serialize state at every epoch and to serialize validators to reload state. This is mainly to avoid memory allocation, plus we may have an optimized way to serialize validators as in `serializeState.test.ts`